src/lib.rs

```diff
@@ -416,65 +416,59 @@ impl NeuQuant {
-                } else {
-                    e = p.b - b as i32;
-                    dist += e * e;
-                    if dist < bestd {
-                        e = p.r - r as i32;
-                        dist += e * e;
-                        if dist < bestd {
-                            e = p.a - a as i32;
-                            dist += e * e;
-                            if dist < bestd {
-                                bestd = dist;
-                                best = i;
-                            }
-                        }
-                    }
-                    i += 1;
+        let mut best_dist = std::i32::MAX;
+        let first_guess = self.netindex[g as usize];
+        let mut best_pos = first_guess;
+        let mut i = best_pos;
+
+        #[inline]
+        fn sqr_dist(a: i32, b: u8) -> i32 {
+            let dist = a - b as i32;
+            dist * dist
+        }
+
+        {
+            let mut cmp = |i| {
+                let Quad {
+                    r: pr,
+                    g: pg,
+                    b: pb,
+                    a: pa,
+                } = self.colormap[i];
+                let mut dist = sqr_dist(pg, g);
+                if dist > best_dist {
+                    return true;
                 }
-            }
-            if j > 0 {
-                let p = self.colormap[j];
-                let mut e = p.g - g as i32;
-                let mut dist = e * e; // inx key
-                if dist >= bestd {
-                    break;
-                } else {
-                    e = p.b - b as i32;
-                    dist += e * e;
-                    if dist < bestd {
-                        e = p.r - r as i32;
-                        dist += e * e;
-                        if dist < bestd {
-                            e = p.a - a as i32;
-                            dist += e * e;
-                            if dist < bestd {
-                                bestd = dist;
-                                best = j;
-                            }
-                        }
-                    }
-                    j -= 1;
+                dist += sqr_dist(pr, r);
+                if dist >= best_dist {
+                    return false;
+                }
+                dist += sqr_dist(pb, b);
+                if dist >= best_dist {
+                    return false;
                 }
+                dist += sqr_dist(pa, a);
+                if dist >= best_dist {
+                    return false;
+                }
+                best_dist = dist;
+                best_pos = i;
+                false
+            };
+            while i < self.netsize {
+                i = if cmp(i) { break } else { i + 1 };
+            }
+            // this j < C is a cheat to avoid the bounds check, as when the loop reaches 0
+            // it will wrap to usize::MAX. Assume that self.netsize < usize::MAX, otherwise
+            // using a lot of memory.
```

Comment:
Neat. That's even a hard guarantee and it's impossible for this to fail, though that is far from obvious from this local code. We allocate multiple vectors of length `self.netsize` and LLVM and Rust both require that allocations have a size of at most `isize::MAX`. So `self.netsize < isize::MAX < usize::MAX` must hold when we reach this point.